### PR TITLE
Fix for switching active queries using file explorer

### PR DESCRIPTION
**Pull request by smartguest**
See associated issue for details:

When clicking the file entry in file explorer, it causes the connection to reconnect with the file input and open in the same editor window as the previous connection to the file. During this phase, the current database name will be in the process of updating itself, and what happens is that the query editor upon having the input state being connected, will immediately try to grab the current database when it isn't reassigned fully yet and thus the resulting database is undefined.  

To fix this issue, we need to have a delay to wait for the database name to fully update itself before we can continue, giving at least 200 MS before retrieving it.

It may be possible to switch to the existing input and/or connectionProfile when clicking file explorer, but this may result in complications with connection callbacks not working properly and the input may not be up to date with the current file, so this is a complicated and not practical solution.

this PR fixes: https://github.com/microsoft/azuredatastudio/issues/16992

### CODE REVIEW

**Comment by smartguest** [comment]
Will need to fix the tests before this PR is good to go.

File: src/sql/workbench/contrib/query/browser/queryActions.ts [hunk 1]
```diff
@@ -640,8 +640,11 @@ export class ListDatabasesActionItem extends Disposable implements IActionViewIt
 
 	// EVENT HANDLERS FROM EDITOR //////////////////////////////////////////
 	public onConnected(): void {
-		let dbName = this.getCurrentDatabaseName();
-		this.updateConnection(dbName);
+		// Need to delay in order to wait for the current database to reassign itself if reconnecting via file explorer.
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/query/browser/queryActions.ts:643]: And how do we know 200ms is enough? And what does it mean for the current database to "reassign itself"? We should only be getting this event when it's actually connected so if that isn't happening that seems like the bug we should be fixing instead.

Why is it creating a new connection when selecting from the file explorer anyways? If the file already has an editor opened then we should be finding that one and use the existing connection.

**Review by Charles-Gagnon — CHANGES_REQUESTED** [review]

File: src/sql/workbench/contrib/query/browser/queryActions.ts (hunk 1 above)
Comment (smartguest) [reply src/sql/workbench/contrib/query/browser/queryActions.ts:643]: I got the value from 200ms after doing tests with different values. 

What I mean by reassign itself is that when the function calls getCurrentDatabaseName, it tries to call getConnectionProfile on the connectionManagementService, which returns undefined (this behavior is only noticed when not debugging before that line, so this is a race condition)

syncConvertInput creates a new EditorInput when the file editor input is selected, and then calls connect with the profile that is retrieved, with new connectioncompletionoptions, this calls tryConnect, which then calls createNewConnection

**Comment by smartguest** [comment]
Apparently checking for an existing connection profile and not creating a new one works fine, the input is updated appropriately and it does not affect the status of the connection profile.

File: src/sql/workbench/contrib/query/browser/queryInputFactory.ts [hunk 2]
```diff
@@ -83,7 +83,9 @@ export class QueryEditorLanguageAssociation implements ILanguageAssociation {
 		}
 
 		const profile = getCurrentGlobalConnection(this.objectExplorerService, this.connectionManagementService, this.editorService);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/query/browser/queryInputFactory.ts:85]: Make the existing profile check into an outer if and then only get the global connection if we don't have an existing one currently.

Make sure you thoroughly test opening queries as well - especially when untitled queries are re-used (new untitled query -> change connection to a different db -> close query window -> new untitled query again which should re-use the same name)

Comment (smartguest) [reply src/sql/workbench/contrib/query/browser/queryInputFactory.ts:85]: It doesn't seem to call the syncConvertInput neither is the convertInput whatsoever, so this does not affect untitledQueries it seems

Comment (smartguest) [reply src/sql/workbench/contrib/query/browser/queryInputFactory.ts:85]: also convertInput is never called, so I'm not sure if I should add the existing check for it as well, wouldn't hurt though.

**Comment by smartguest** [comment]
Will need to fix the tests as the global connection not being created if an existing connection profile is found.

File: src/sql/workbench/contrib/query/browser/queryInputFactory.ts [hunk 3]
```diff
@@ -82,16 +87,20 @@ export class QueryEditorLanguageAssociation implements ILanguageAssociation {
...
-		if (profile) {
-			const options: IConnectionCompletionOptions = {
-				params: { connectionType: ConnectionType.editor, runQueryOnCompletion: undefined, input: queryEditorInput },
-				saveTheConnection: false,
-				showDashboard: false,
-				showConnectionDialogOnError: true,
-				showFirewallRuleOnError: true
-			};
-			this.connectionManagementService.connect(profile, queryEditorInput.uri, options).catch(err => onUnexpectedError(err));
+		const existingProfile = this.connectionManagementService.getConnectionProfile(queryEditorInput.uri);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/query/browser/queryInputFactory.ts:90]: Would be good to just put these both in a helper function since they seem to be doing the same thing

File: src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts [hunk 4]
```diff
@@ -141,6 +141,217 @@ suite('Query Input Factory', () => {
...
+		instantiationService.stub(IConnectionManagementService, connectionManagementService);
+		instantiationService.stub(IEditorService, editorService);
+		const queryEditorLanguageAssociation = instantiationService.createInstance(QueryEditorLanguageAssociation);
+		const input1 = createFileInput(URI.file('/test/file.sql'), undefined, undefined, undefined);
+		const input2 = createFileInput(URI.file('/test/file.sql'), undefined, undefined, undefined);
+		queryEditorLanguageAssociation.syncConvertInput(input1);
+		queryEditorLanguageAssociation.syncConvertInput(input2);
+		let connProfile = connectionManagementService.getConnectionProfile('file:///test/file.sql');
+		assert(connProfile !== undefined, 'connection profile should not be undefined');
+		assert(connProfile.azureResourceId === 'file:///test/file.sql', 'azureResourceId should not be undefined');
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts:158]: Why are you testing the azureResourceId? And why is it being set to the file URI? That doesn't seem right - the resource ID should correspond to a value from this enum : https://github.com/microsoft/azuredatastudio/blob/main/src/sql/azdata.d.ts#L2277

File: src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts [hunk 5]
```diff
@@ -141,6 +141,217 @@ suite('Query Input Factory', () => {
...
 
+	test('sync query editor input can only be connected once when global connection exists', () => {
+		const instantiationService = workbenchInstantiationService();
+		const editorService = new MockEditorService();
+		const connectionManagementService = new MockConnectionManagementService();
+		instantiationService.stub(IObjectExplorerService, new MockObjectExplorerService());
+		instantiationService.stub(IConnectionManagementService, connectionManagementService);
+		instantiationService.stub(IEditorService, editorService);
+		const queryEditorLanguageAssociation = instantiationService.createInstance(QueryEditorLanguageAssociation);
+		const input1 = createFileInput(URI.file('/test/file.sql'), undefined, undefined, undefined);
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts:152]: For these tests I think you should make it more clear that you're testing creating multiple inputs with the same file URI - in the title of the test, the comments and the error message verifying the number of connects.

File: src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts [hunk 6]
```diff
@@ -141,6 +141,217 @@ suite('Query Input Factory', () => {
...
+		let connProfile1 = connectionManagementService.getConnectionProfile('file:///test/file1.sql');
+		assert(connProfile1 !== undefined, 'connection profile should not be undefined');
+		assert(connProfile1.azureResourceId === 'file:///test/file1.sql', 'azureResourceId should not be undefined');
+		let connProfile2 = connectionManagementService.getConnectionProfile('file:///test/file2.sql');
+		assert(connProfile2 !== undefined, 'connection profile should not be undefined');
+		assert(connProfile2.azureResourceId === 'file:///test/file2.sql', 'azureResourceId should not be undefined');
+		assert(connectionManagementService.numberConnects === 2, 'Convert input should have called connect two times when active OE connection exists and we have two inputs');
+	});
+
+	test('query editor input can only be connected once when global connection exists (OE)', async () => {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts:183]: What is the OE part referring to? As far as I can tell this test is doing the same thing as the others (it's just using convertInput instead of syncConvertInput)

Same with the "Editor" tests below

File: src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts [hunk 7]
```diff
@@ -205,16 +416,25 @@ class MockConnectionManagementService extends TestConnectionManagementService {
...
 		this.numberConnects++;
+		this.connectionProfiles.push(uri);
 		return Promise.resolve(undefined);
 	}
 
 	public override getConnectionProfile(fileUri: string): IConnectionProfile {
-		return <IConnectionProfile>{}; // Not actually used so fine to cast
+		let element = this.connectionProfiles.find(element => fileUri === element);
+		if (element) {
+			return <IConnectionProfile>{ azureResourceId: element };
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts:434]: Yeah, not really sure what this is supposed to be doing. This isn't what azureResourceId is supposed to have. 

Why aren't you caching the actual profiles passed in during connect and then returning those instead?

File: src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts [hunk 8]
```diff
@@ -141,6 +141,217 @@ suite('Query Input Factory', () => {
 		assert(connectionManagementService.numberConnects === 0, 'Convert input should not have been called connect when no global connections exist');
 	});
 
+	test('sync query editor input can only be connected once when global connection exists', () => {
```
Comment (Charles-Gagnon) [inline src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts:144]: What about tests when a global connection doesn't exist?

File: src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts (hunk 6 above)
Comment (smartguest) [reply src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts:183]: Just following what the other tests were doing above, which I don't know why that's the case.

Comment (smartguest) [reply src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts:183]: Only difference between the Editor and Object Explorer tests are that the Object Explorer tests use a mock Object Explorer Service Test, while the Editor ones do not.

File: src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts (hunk 7 above)
Comment (smartguest) [reply src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts:434]: Doing that right now

File: src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts (hunk 4 above)
Comment (smartguest) [reply src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts:158]: Changing it to return actual profile

File: src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts (hunk 8 above)
Comment (smartguest) [reply src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts:144]: Those tests already exist in this file.

**Comment by smartguest** [comment]
I've moved the tests around so that they follow in a more logical order, they simulate running under object explorer and editor respectively with the main difference being the MockEditorService being able to assign an active editor with untitledInput.

**Review by Charles-Gagnon — APPROVED** [review]
let's make sure that this stuff is heavily tested for the next release, there's a lot of potential for edge case regressions here with regards to connections, results and how they inputs have those set.

File: src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts [hunk 9]
```diff
@@ -112,7 +195,90 @@ suite('Query Input Factory', () => {
...
+		const connectionManagementService = new MockConnectionManagementService();
+		instantiationService.stub(IObjectExplorerService, new MockObjectExplorerService());
+		instantiationService.stub(IConnectionManagementService, connectionManagementService);
+		instantiationService.stub(IEditorService, editorService);
+		const queryEditorLanguageAssociation = instantiationService.createInstance(QueryEditorLanguageAssociation);
+		const input1 = createFileInput(URI.file('/test/file1.sql'), undefined, undefined, undefined);
+		const input2 = createFileInput(URI.file('/test/file2.sql'), undefined, undefined, undefined);
+		queryEditorLanguageAssociation.syncConvertInput(input1);
+		queryEditorLanguageAssociation.syncConvertInput(input2);
+		let connProfile1 = connectionManagementService.getConnectionProfile('file:///test/file1.sql');
```
Comment (alanrenmsft) [inline src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts:230]: have you run these tests on both windows and mac at least? the path resolution might have some issues here

**Review by alanrenmsft — APPROVED** [review]

File: src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts (hunk 9 above)
Comment (smartguest) [reply src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts:230]: Tests were run on Windows just fine, Mac is fine according to CI, I am going to investigate on my Mac itself if that is needed

Comment (smartguest) [reply src/sql/workbench/contrib/query/test/browser/queryInputFactory.test.ts:230]: Works fine on my actual Mac, so far, seems good.
